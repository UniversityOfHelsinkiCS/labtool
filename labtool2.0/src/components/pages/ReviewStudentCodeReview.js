import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router'
import { Button, Form, Input, Grid, Card, Loader, Icon } from 'semantic-ui-react'
import { Link, Redirect } from 'react-router-dom'
import { connect } from 'react-redux'
import { getOneCI, coursePageInformation } from '../../services/courseInstance'
import { clearNotifications } from '../../reducers/notificationReducer'
import { gradeCodeReview } from '../../services/codeReview'
import { toggleCheckCodeReview, resetChecklist, restoreChecks } from '../../reducers/weekReviewReducer'
import { resetLoading, addRedirectHook } from '../../reducers/loadingReducer'
import store from '../../store'
import { usePersistedState } from '../../hooks/persistedState'

import RepoLink from '../RepoLink'
import BackButton from '../BackButton'

/**
 *  The page which is used by teacher to review submissions,.
 */
export const ReviewStudentCodeReview = props => {
  const [loadedCrData, setLoadedCrData] = useState(false)
  const [allowChecksCopy, setAllowChecksCopy] = useState(false)
  const pstate = usePersistedState(`ReviewStudentCodeReview_${props.studentInstance}:${props.codeReviewNumber}`, {
    points: '',
    checks: null
  })

  useEffect(() => {
    // run on component mount
    props.resetLoading()
    props.getOneCI(props.courseId)
    props.coursePageInformation(props.courseId)
    props.clearNotifications()

    return () => {
      // run on component unmount
      props.resetChecklist()
    }
  }, [])

  useEffect(() => {
    if (allowChecksCopy && props.weekReview.checks !== null) {
      const d = { ...props.weekReview.checks }
      pstate.checks = d
    }
  }, [props.weekReview.checks])

  const constructChecks = thisChecks => {
    let checksObj = {}
    Object.keys(checkList.list).forEach(category => {
      checkList.list[category].forEach(clItem => {
        checksObj[clItem.id] = false
      })
    })
    checksObj = { ...checksObj, ...thisChecks }
    return Object.keys(checksObj).map(id => ({ id: Number(id, 10), checked: checksObj[id] }))
  }

  const decodeChecks = checksList => {
    const checksObj = {}
    checksList.forEach(({ checklistItemId, checked }) => (checksObj[checklistItemId] = checked))
    return checksObj
  }

  const handleSubmit = async e => {
    try {
      e.preventDefault()
      const content = {
        points: Number(e.target.points.value),
        checks: constructChecks(props.weekReview.checks),
        studentInstanceId: Number(props.studentInstance),
        reviewNumber: Number(props.codeReviewNumber)
      }
      pstate.clear()
      const maxPoints = getMaximumPoints()
      if (e.target.points.value < 0 || (maxPoints && e.target.points.value > maxPoints)) {
        store.dispatch({ type: 'CODE_REVIEW_GRADE_FAILURE', response: { response: { data: 'Invalid points value' } } })
      } else {
        props.addRedirectHook({
          hook: 'CODE_REVIEW_GRADE_'
        })
        await props.gradeCodeReview(content)
      }
    } catch (error) {
      console.error(error)
    }
  }

  const getMaximumPoints = () => {
    const checklist = props.selectedInstance.checklists.find(checkl => checkl.codeReviewNumber === props.codeReviewNumber)
    if (checklist && checklist.maxPoints) {
      return checklist.maxPoints
    }
    return null
  }

  const toggleCheckbox = (checklistItemId, studentId, crNbr) => async () => {
    setAllowChecksCopy(true)
    props.toggleCheckCodeReview(checklistItemId, studentId, crNbr)
  }

  const copyChecklistOutput = async e => {
    e.preventDefault()
    pstate.points = e.target.points.value
  }

  const isChecked = (checks, checklistItemId) =>
    checks !== null && checks[checklistItemId] !== undefined
      ? checks[checklistItemId]
      : props.weekReview.checks !== null && props.weekReview.checks[checklistItemId] !== undefined
      ? props.weekReview.checks[checklistItemId]
      : false

  if (props.loading.loading) {
    return <Loader active />
  }
  if (props.loading.redirect) {
    pstate.clear()
    return <Redirect to={`/labtool/courses/${props.selectedInstance.ohid}`} />
  }
  if (!Array.isArray(props.weekReview.data)) {
    return <Loader active />
  }

  const student = props.courseData.data.find(student => student.id === Number(props.studentInstance))
  const cr = student.codeReviews.find(cr => cr.reviewNumber === Number(props.codeReviewNumber))
  const checkList = props.selectedInstance.checklists.find(checkl => checkl.codeReviewNumber === Number(props.codeReviewNumber))
  const maxPoints = getMaximumPoints()
  const weekPoints = student.weeks
    .map(week => week.points)
    .reduce((a, b) => {
      return a + b
    }, 0)
  const codeReviewPoints = student.codeReviews
    .filter(review => review.reviewNumber < Number(props.codeReviewNumber))
    .map(review => review.points)
    .reduce((a, b) => {
      return a + b
    }, 0)
  const savedChecks = cr ? cr.checks || {} : {}
  const checks = props.weekReview.checks !== null ? props.weekReview.checks : savedChecks
  let checklistPoints = 0
  const toReviewProject = cr.toReview ? props.courseData.data.find(student => student.id === cr.toReview) : null
  if (checkList) {
    Object.keys(checkList.list).forEach(category => {
      checkList.list[category].forEach(clItem => {
        //handle existing case where clItems were saved by name in weekData.checks
        if (savedChecks[clItem.name]) {
          savedChecks[clItem.id] = savedChecks[clItem.name]
          delete savedChecks[clItem.name]
        }

        const checked = isChecked(checks, clItem.id)
        if (checked) {
          checklistPoints += clItem.checkedPoints
        } else {
          checklistPoints += clItem.uncheckedPoints
        }
      })
    })
    if (checklistPoints < 0) {
      checklistPoints = 0
    } else if (maxPoints !== null && checklistPoints > maxPoints) {
      checklistPoints = maxPoints
    }
  }

  if (!loadedCrData) {
    if (cr.points) {
      if (cr.checks) {
        props.restoreChecks(props.ownProps.studentInstance, decodeChecks(cr.checks))
      }

      pstate.points = pstate.points || cr.points
      setLoadedCrData(true)
    }
  }

  const arrivedFromCoursePage = props.location && props.location.state && props.location.state.cameFromCoursePage

  return (
    <div className="ReviewStudentCodeReview">
      <BackButton
        preset={arrivedFromCoursePage && 'coursePage'}
        to={!arrivedFromCoursePage && `/labtool/browsereviews/${props.selectedInstance.ohid}/${student.id}`}
        text={!arrivedFromCoursePage && 'Back to student reviews'}
      />
      <div style={{ textAlignVertical: 'center', textAlign: 'center' }}>
        <h2> {props.selectedInstance.name}</h2>
        <h3>
          {' '}
          {student.User.firsts} {student.User.lastname}
        </h3>
        <h3>Code Review {props.codeReviewNumber}</h3>
        <Grid>
          <Grid.Row columns={2}>
            <Grid.Column>
              <div align="left">
                <h3>Points so far: {weekPoints + codeReviewPoints} </h3>
                Earlier code review points: {codeReviewPoints}
                <br />
                Week points: {weekPoints}
              </div>
              <h3>Project to review</h3>
              {toReviewProject ? (
                <p>
                  {toReviewProject.projectName}: <RepoLink url={toReviewProject.github} />
                </p>
              ) : cr.repoToReview ? (
                <RepoLink url={cr.repoToReview} />
              ) : (
                <p>No assigned project to review</p>
              )}
              <h3>Link to code review</h3>
              {cr.linkToReview ? (
                <p>
                  <a target="_blank" rel="noopener noreferrer" href={cr.linkToReview}>
                    {cr.linkToReview}
                  </a>
                </p>
              ) : (
                <p>No code review linked</p>
              )}
              <h2>Review</h2>
              <Form onSubmit={handleSubmit}>
                <Form.Group inline unstackable>
                  <Form.Field>
                    <label className="showMaxPoints">Points{maxPoints !== null ? `0-${maxPoints}` : ''}</label>

                    <Input
                      name="points"
                      required={true}
                      value={pstate.points}
                      onChange={(e, { value }) => (pstate.points = value)}
                      type="number"
                      step="0.01"
                      style={{ width: '150px', align: 'center' }}
                    />
                  </Form.Field>
                </Form.Group>
                <Form.Field>
                  <Button className="ui center floated green button" type="submit">
                    Save
                  </Button>
                  <Link to={`/labtool/browsereviews/${props.selectedInstance.ohid}/${student.id}`} type="Cancel">
                    <Button className="ui center floated button" type="cancel" onClick={pstate.clear}>
                      Cancel
                    </Button>
                  </Link>
                </Form.Field>
              </Form>
            </Grid.Column>
            {checkList && checks !== undefined ? (
              <Grid.Column>
                <h2>Checklist</h2>
                {checkList ? (
                  <div className="checklist">
                    {Object.keys(checkList.list).map(clItemCategory => (
                      <Card className="checklistCard" fluid color="red" key={clItemCategory}>
                        <Card.Content header={clItemCategory} />
                        {checkList.list[clItemCategory].map(clItem => (
                          <Card.Content className="checklistCardRow" key={clItem.id} onClick={toggleCheckbox(clItem.id, props.ownProps.studentInstance, props.ownProps.weekNumber)}>
                            <Form.Field>
                              <Grid>
                                <Grid.Row style={{ cursor: 'pointer', userSelect: 'none' }}>
                                  <Grid.Column width={3}>
                                    <Icon
                                      size="large"
                                      name={isChecked(checks, clItem.id) ? 'circle check outline' : 'circle outline'}
                                      style={{ color: isChecked(checks, clItem.id) ? 'green' : 'black' }}
                                    />
                                  </Grid.Column>
                                  <Grid.Column width={10}>
                                    <span style={{ flexGrow: 1, textAlign: 'center' }}>{clItem.name}</span>
                                  </Grid.Column>
                                  <Grid.Column width={3}>
                                    <span>{`${clItem.checkedPoints} p / ${clItem.uncheckedPoints} p`}</span>
                                  </Grid.Column>
                                </Grid.Row>
                              </Grid>
                            </Form.Field>
                          </Card.Content>
                        ))}
                      </Card>
                    ))}
                    <div>
                      <Form className="checklistOutput" onSubmit={copyChecklistOutput}>
                        <p className="checklistOutputPoints">points: {checklistPoints.toFixed(2)}</p>
                        <input type="hidden" name="points" value={checklistPoints} />
                        <Button type="submit">Copy to review field</Button>
                      </Form>
                    </div>
                  </div>
                ) : (
                  <p>There is no checklist for this code review.</p>
                )}
              </Grid.Column>
            ) : (
              <div />
            )}
          </Grid.Row>
        </Grid>
      </div>
    </div>
  )
}

const mapStateToProps = (state, ownProps) => {
  return {
    ownProps,
    selectedInstance: state.selectedInstance,
    courseData: state.coursePage,
    weekReview: state.weekReview,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  getOneCI,
  clearNotifications,
  gradeCodeReview,
  toggleCheckCodeReview,
  restoreChecks,
  resetChecklist,
  coursePageInformation,
  resetLoading,
  addRedirectHook
}

ReviewStudentCodeReview.propTypes = {
  ownProps: PropTypes.object.isRequired,

  courseId: PropTypes.string.isRequired,
  studentInstance: PropTypes.string.isRequired,
  codeReviewNumber: PropTypes.string.isRequired,

  selectedInstance: PropTypes.object.isRequired,
  courseData: PropTypes.object.isRequired,
  weekReview: PropTypes.object.isRequired,
  loading: PropTypes.object.isRequired,
  location: PropTypes.object,

  getOneCI: PropTypes.func.isRequired,
  clearNotifications: PropTypes.func.isRequired,
  gradeCodeReview: PropTypes.func.isRequired,
  toggleCheckCodeReview: PropTypes.func.isRequired,
  restoreChecks: PropTypes.func.isRequired,
  resetChecklist: PropTypes.func.isRequired,
  coursePageInformation: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  addRedirectHook: PropTypes.func.isRequired
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ReviewStudentCodeReview)
)
