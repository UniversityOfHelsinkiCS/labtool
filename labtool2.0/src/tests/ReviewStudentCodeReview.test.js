import React from 'react'
import { ReviewStudentCodeReview } from '../components/pages/ReviewStudentCodeReview'
import { shallow } from 'enzyme'

describe('<ReviewStudentCodeReviews />', () => {
  let wrapper

  const ownProps = {
    courseId: 'TKT20010.2018.K.A.1',
    codeReviewNumber: '1',
    studentInstance: '10011'
  }

  const props = {
    selectedInstance: {
      id: 10011,
      name: 'Aineopintojen harjoitustyö: Tietorakenteet ja algoritmit',
      start: '2018-03-11T21:00:00.000Z',
      end: '2018-04-29T21:00:00.000Z',
      active: true,
      weekAmount: 7,
      weekMaxPoints: 3,
      currentWeek: 1,
      ohid: 'TKT20010.2018.K.A.1',
      checklists: [
        {
          id: 20003,
          codeReviewNumber: 1,
          list: {
            Koodi: [
              {
                id: 1,
                name: 'Koodin laatu',
                points: 2,
                textWhenOn: 'Koodi tehty laadukkaasti ja sisältää kommentteja',
                textWhenOff: 'Koodin laadussa parantamisen varaa'
              }
            ],
            Algoritmit: [
              {
                id: 2,
                name: 'Algoritmin runko',
                points: 2,
                textWhenOn: 'Algoritmin runko luotu',
                textWhenOff: 'Algoritmin runko puuttuu'
              },
              {
                id: 3,
                name: 'Tietorakenteita luotu',
                points: 2,
                textWhenOn: 'Tietorakenteita luotu',
                textWhenOff: 'Tietorakenteita ei ole luotu'
              }
            ],
            Dokumentaatio: [
              {
                id: 4,
                name: 'Readme',
                points: 1,
                textWhenOn: 'README kunnossa',
                textWhenOff: 'README puuttuu'
              },
              {
                id: 5,
                name: 'Tuntikirjanpito',
                points: 1,
                textWhenOn: 'Tuntikirjanpito täytetty oikein',
                textWhenOff: 'Tuntikirjanpito puuttuu'
              }
            ]
          },
          courseName: 'Aineopintojen harjoitustyö: Tietorakenteet ja algoritmit',
          master: false,
          maxPoints: 4,
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          courseInstanceId: 10011
        }
      ],
      createdAt: '2018-03-26T00:00:00.000Z',
      updatedAt: '2018-03-26T00:00:00.000Z',
      teacherInstances: [
        {
          id: 10001,
          instructor: false,
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          userId: 10010,
          courseInstanceId: 10011,
          firsts: 'Pää',
          lastname: 'Opettaja'
        },
        {
          id: 10011,
          instructor: true,
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          userId: 10015,
          courseInstanceId: 10011,
          firsts: 'Ossi Ohjaaja',
          lastname: 'Mutikainen'
        }
      ]
    },
    studentInstance: 10011,
    coursePage: {
      role: 'teacher',
      data: [
        {
          id: 10012,
          github: 'http://github.com/tiralabra2',
          projectName: 'Tiran toinen labraprojekti',
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          courseInstanceId: 10011,
          userId: 10012,
          teacherInstanceId: 10011,
          weeks: [],
          codeReviews: [],
          User: {
            id: 10012,
            username: 'tiraopiskelija2',
            email: 'johan.studerande@helsinki.invalid',
            firsts: 'Johan Wilhelm',
            lastname: 'Studerande',
            studentNumber: '014553242',
            teacher: false,
            sysop: false,
            createdAt: '2018-03-26T00:00:00.000Z',
            updatedAt: '2018-03-26T00:00:00.000Z'
          }
        },
        {
          id: 10031,
          github: 'http://github.com/superprojekti',
          projectName: 'Tira super projekti',
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-06-05T07:12:28.603Z',
          courseInstanceId: 10011,
          userId: 10031,
          teacherInstanceId: 10011,
          weeks: [],
          codeReviews: [],
          User: {
            id: 10031,
            username: 'superopiskelija',
            email: 'teras.henkilo@helsinki.invalid',
            firsts: 'Teräs',
            lastname: 'Henkilö',
            studentNumber: '014666666',
            teacher: false,
            sysop: false,
            createdAt: '2018-03-26T00:00:00.000Z',
            updatedAt: '2018-03-26T00:00:00.000Z'
          }
        },
        {
          id: 10011,
          github: 'http://github.com/tiralabra1',
          projectName: 'Tiran labraprojekti',
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          courseInstanceId: 10011,
          userId: 10011,
          teacherInstanceId: 10011,
          weeks: [
            {
              id: 10002,
              points: 2,
              weekNumber: 2,
              feedback: 'Melko hienoa työtä!',
              createdAt: '2018-03-26T00:00:00.000Z',
              updatedAt: '2018-03-26T00:00:00.000Z',
              studentInstanceId: 10011,
              comments: []
            },
            {
              id: 10001,
              points: 3,
              weekNumber: 1,
              feedback: 'Hienoa työtä!',
              createdAt: '2018-03-26T00:00:00.000Z',
              updatedAt: '2018-03-26T00:00:00.000Z',
              studentInstanceId: 10011,
              comments: []
            },
            {
              id: 10003,
              points: 3,
              weekNumber: 3,
              feedback: 'Erittäin hienoa työtä!',
              createdAt: '2018-03-26T00:00:00.000Z',
              updatedAt: '2018-03-26T00:00:00.000Z',
              studentInstanceId: 10011,
              comments: []
            },
            {
              id: 10004,
              points: 3,
              weekNumber: 4,
              feedback: 'Hyvin menee!',
              createdAt: '2018-03-26T00:00:00.000Z',
              updatedAt: '2018-03-26T00:00:00.000Z',
              studentInstanceId: 10011,
              comments: []
            }
          ],
          codeReviews: [
            {
              id: 10001,
              points: null,
              reviewNumber: 1,
              linkToReview: 'http://github.com/tiralabra2/issues/1',
              repoToReview: 'http://github.com/tiralabra2',
              createdAt: '2018-03-26T00:00:00.000Z',
              updatedAt: '2018-03-26T00:00:00.000Z',
              studentInstanceId: 10011
            }
          ],
          User: {
            id: 10011,
            username: 'tiraopiskelija1',
            email: 'maarit.opiskelija@helsinki.invalid',
            firsts: 'Maarit Mirja',
            lastname: 'Opiskelija',
            studentNumber: '014578343',
            teacher: false,
            sysop: false,
            createdAt: '2018-03-26T00:00:00.000Z',
            updatedAt: '2018-03-26T00:00:00.000Z'
          }
        }
      ]
    },
    weekReview: {
      data: [
        {
          id: 10031,
          github: 'http://github.com/superprojekti',
          projectName: 'Tira super projekti',
          courseInstanceId: 10011,
          userId: 10031,
          teacherInstanceId: 10001,
          weeks: [
            {
              id: 1,
              points: 0,
              weekNumber: 1,
              feedback: 'Koodin laadussa parantamisen varaa\n\nAlgoritmin runko puuttuu\n\nTietorakenteita ei ole luotu\n\nREADME puuttuu\n\nTuntikirjanpito puuttuu\n\n',
              notified: false,
              checks: {
                'Algoritmin runko': true,
                'Tietorakenteita luotu': true,
                'Koodin laatu': true
              },
              studentInstanceId: 10031,
              comments: []
            }
          ],
          codeReviews: [
            {
              id: 10003,
              points: null,
              reviewNumber: 1,
              linkToReview: null,
              studentInstanceId: 10031,
              toReview: 10011
            },
            {
              id: 10008,
              points: null,
              reviewNumber: 2,
              linkToReview: null,
              studentInstanceId: 10031,
              toReview: 10012
            }
          ],
          User: {
            id: 10031,
            username: 'superopiskelija',
            email: 'teras.henkilo@helsinki.invalid',
            firsts: 'Teräs',
            lastname: 'Henkilö',
            studentNumber: '014666666',
            teacher: false,
            sysop: false
          },
          Tags: [
            {
              id: 20007,
              name: 'FORTRAN',
              color: 'pink',
              StudentTag: {
                id: 30007,
                studentInstanceId: 10031,
                tagId: 20007,
                createdAt: '2018-06-13T00:00:00.000Z',
                updatedAt: '2018-06-13T00:00:00.000Z'
              }
            }
          ]
        },
        {
          id: 10011,
          github: 'http://github.com/tiraopi1/tiralabra1',
          projectName: 'Tiran labraprojekti',
          courseInstanceId: 10011,
          userId: 10011,
          teacherInstanceId: 10011,
          weeks: [
            {
              id: 10002,
              points: 2,
              weekNumber: 2,
              feedback: 'Melko hienoa työtä!',
              notified: false,
              checks: {},
              studentInstanceId: 10011,
              comments: []
            },
            {
              id: 10001,
              points: 3,
              weekNumber: 1,
              feedback: 'Hienoa työtä!',
              notified: false,
              checks: {},
              studentInstanceId: 10011,
              comments: []
            },
            {
              id: 10003,
              points: 3,
              weekNumber: 3,
              feedback: 'Erittäin hienoa työtä!',
              notified: false,
              checks: {},
              studentInstanceId: 10011,
              comments: []
            },
            {
              id: 10004,
              points: 3,
              weekNumber: 4,
              feedback: 'Hyvin menee!',
              notified: false,
              checks: {},
              studentInstanceId: 10011,
              comments: []
            }
          ],
          codeReviews: [
            {
              id: 10001,
              points: null,
              reviewNumber: 1,
              linkToReview: null,
              studentInstanceId: 10011,
              toReview: 10012,
              checks: {}
            },
            {
              id: 10007,
              points: null,
              reviewNumber: 2,
              linkToReview: null,
              studentInstanceId: 10011,
              toReview: 10031,
              checks: {}
            }
          ],
          User: {
            id: 10011,
            username: 'tiraopiskelija1',
            email: 'maarit.opiskelija@helsinki.invalid',
            firsts: 'Maarit Mirja',
            lastname: 'Opiskelija',
            studentNumber: '014578343',
            teacher: false,
            sysop: false
          },
          Tags: [
            {
              id: 20003,
              name: 'game',
              color: 'black',
              StudentTag: {
                id: 30003,
                studentInstanceId: 10011,
                tagId: 20003,
                createdAt: '2018-06-13T00:00:00.000Z',
                updatedAt: '2018-06-13T00:00:00.000Z'
              }
            },
            {
              id: 20002,
              name: 'HTML',
              color: 'yellow',
              StudentTag: {
                id: 30002,
                studentInstanceId: 10011,
                tagId: 20002,
                createdAt: '2018-06-13T00:00:00.000Z',
                updatedAt: '2018-06-13T00:00:00.000Z'
              }
            },
            {
              id: 20001,
              name: 'Javascript',
              color: 'red',
              StudentTag: {
                id: 30001,
                studentInstanceId: 10011,
                tagId: 20001,
                createdAt: '2018-06-13T00:00:00.000Z',
                updatedAt: '2018-06-13T00:00:00.000Z'
              }
            }
          ]
        },
        {
          id: 10012,
          github: 'http://github.com/tiraopi2/tiralabra2',
          projectName: 'Tiran toinen labraprojekti',
          courseInstanceId: 10011,
          userId: 10012,
          teacherInstanceId: 10011,
          weeks: [],
          codeReviews: [
            {
              id: 10002,
              points: null,
              reviewNumber: 1,
              linkToReview: null,
              studentInstanceId: 10012,
              toReview: 10031
            },
            {
              id: 10009,
              points: null,
              reviewNumber: 2,
              linkToReview: null,
              studentInstanceId: 10012,
              toReview: 10011
            }
          ],
          User: {
            id: 10012,
            username: 'tiraopiskelija2',
            email: 'johan.studerande@helsinki.invalid',
            firsts: 'Johan Wilhelm',
            lastname: 'Studerande',
            studentNumber: '014553242',
            teacher: false,
            sysop: false
          },
          Tags: [
            {
              id: 20005,
              name: 'Node.js',
              color: 'blue',
              StudentTag: {
                id: 30005,
                studentInstanceId: 10012,
                tagId: 20005,
                createdAt: '2018-06-13T00:00:00.000Z',
                updatedAt: '2018-06-13T00:00:00.000Z'
              }
            },
            {
              id: 20001,
              name: 'Javascript',
              color: 'red',
              StudentTag: {
                id: 30004,
                studentInstanceId: 10012,
                tagId: 20001,
                createdAt: '2018-06-13T00:00:00.000Z',
                updatedAt: '2018-06-13T00:00:00.000Z'
              }
            }
          ]
        }
      ],
      checks: {}
    },
    loading: {
      loading: false,
      loadingHooks: [],
      redirect: false,
      redirectHooks: [],
      redirectFailure: false
    }
  }

  let mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <ReviewStudentCodeReview
        getOneCI={mockFn}
        clearNotifications={mockFn}
        courseData={props.coursePage}
        ownProps={ownProps}
        selectedInstance={props.selectedInstance}
        weekReview={props.weekReview}
        loading={props.loading}
        resetLoading={mockFn}
        coursePageInformation={mockFn}
        {...ownProps}
        notification={{}}
        toggleCheckCodeReview={mockFn}
        initChecks={mockFn}
        restoreChecks={mockFn}
        resetChecklist={mockFn}
        addRedirectHook={mockFn}
        gradeCodeReview={mockFn}
      />
    )
  })

  describe('ReviewStudentCodeReview Component', () => {
    it('is ok', () => {
      true
    })

    it('should render without throwing an error', () => {
      expect(wrapper.find('.ReviewStudentCodeReview').exists()).toEqual(true)
    })

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
