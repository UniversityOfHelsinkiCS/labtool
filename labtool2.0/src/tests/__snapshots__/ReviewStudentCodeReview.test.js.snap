// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewStudentCodeReviews /> ReviewStudentCodeReview Component should render correctly 1`] = `
<div
  className="ReviewStudentCodeReview"
>
  <Connect(BackButton)
    text="Back to student reviews"
    to="/labtool/browsereviews/TKT20010.2018.K.A.1/10011"
  />
  <div
    style={
      Object {
        "textAlign": "center",
        "textAlignVertical": "center",
      }
    }
  >
    <h2>
       
      Aineopintojen harjoitusty√∂: Tietorakenteet ja algoritmit
    </h2>
    <h3>
       
      Maarit Mirja
       
      Opiskelija
    </h3>
    <h3>
      Code Review 
      1
    </h3>
    <Grid>
      <GridRow
        columns={2}
      >
        <GridColumn>
          <div
            align="left"
          >
            <h3>
              Points so far: 
              11
               
            </h3>
            Earlier code review points: 
            0
            <br />
            Week points: 
            11
          </div>
          <h3>
            Project to review
          </h3>
          <RepoLink
            url="http://github.com/tiralabra2"
          />
          <h3>
            Link to code review
          </h3>
          <p>
            <a
              href="http://github.com/tiralabra2/issues/1"
              rel="noopener noreferrer"
              target="_blank"
            >
              http://github.com/tiralabra2/issues/1
            </a>
          </p>
          <h2>
            Review
          </h2>
          <Form
            as="form"
            onSubmit={[Function]}
          >
            <FormGroup
              inline={true}
              unstackable={true}
            >
              <FormField>
                <label
                  className="showMaxPoints"
                >
                  Points
                </label>
                <Input
                  name="points"
                  onChange={[Function]}
                  required={true}
                  step="0.01"
                  style={
                    Object {
                      "align": "center",
                      "width": "150px",
                    }
                  }
                  type="number"
                  value=""
                />
              </FormField>
            </FormGroup>
            <FormField>
              <Button
                as="button"
                className="ui center floated green button"
                type="submit"
              >
                Save
              </Button>
              <Link
                to="/labtool/browsereviews/TKT20010.2018.K.A.1/10011"
                type="Cancel"
              >
                <Button
                  as="button"
                  className="ui center floated button"
                  onClick={[Function]}
                  type="cancel"
                >
                  Cancel
                </Button>
              </Link>
            </FormField>
          </Form>
        </GridColumn>
        <GridColumn>
          <h2>
            Checklist
          </h2>
          <div
            className="checklist"
          >
            <Card
              className="checklistCard"
              color="red"
              fluid={true}
              key="Koodi"
            >
              <CardContent
                header="Koodi"
              />
              <CardContent
                className="checklistCardRow"
                key="1"
                onClick={[Function]}
              >
                <FormField>
                  <Grid>
                    <GridRow
                      style={
                        Object {
                          "cursor": "pointer",
                          "userSelect": "none",
                        }
                      }
                    >
                      <GridColumn
                        width={3}
                      >
                        <Icon
                          as="i"
                          name="circle outline"
                          size="large"
                          style={
                            Object {
                              "color": "black",
                            }
                          }
                        />
                      </GridColumn>
                      <GridColumn
                        width={10}
                      >
                        <span
                          style={
                            Object {
                              "flexGrow": 1,
                              "textAlign": "center",
                            }
                          }
                        >
                          Koodin laatu
                        </span>
                      </GridColumn>
                      <GridColumn
                        width={3}
                      >
                        <span>
                          undefined p / undefined p
                        </span>
                      </GridColumn>
                    </GridRow>
                  </Grid>
                </FormField>
              </CardContent>
            </Card>
            <Card
              className="checklistCard"
              color="red"
              fluid={true}
              key="Algoritmit"
            >
              <CardContent
                header="Algoritmit"
              />
              <CardContent
                className="checklistCardRow"
                key="2"
                onClick={[Function]}
              >
                <FormField>
                  <Grid>
                    <GridRow
                      style={
                        Object {
                          "cursor": "pointer",
                          "userSelect": "none",
                        }
                      }
                    >
                      <GridColumn
                        width={3}
                      >
                        <Icon
                          as="i"
                          name="circle outline"
                          size="large"
                          style={
                            Object {
                              "color": "black",
                            }
                          }
                        />
                      </GridColumn>
                      <GridColumn
                        width={10}
                      >
                        <span
                          style={
                            Object {
                              "flexGrow": 1,
                              "textAlign": "center",
                            }
                          }
                        >
                          Algoritmin runko
                        </span>
                      </GridColumn>
                      <GridColumn
                        width={3}
                      >
                        <span>
                          undefined p / undefined p
                        </span>
                      </GridColumn>
                    </GridRow>
                  </Grid>
                </FormField>
              </CardContent>
              <CardContent
                className="checklistCardRow"
                key="3"
                onClick={[Function]}
              >
                <FormField>
                  <Grid>
                    <GridRow
                      style={
                        Object {
                          "cursor": "pointer",
                          "userSelect": "none",
                        }
                      }
                    >
                      <GridColumn
                        width={3}
                      >
                        <Icon
                          as="i"
                          name="circle outline"
                          size="large"
                          style={
                            Object {
                              "color": "black",
                            }
                          }
                        />
                      </GridColumn>
                      <GridColumn
                        width={10}
                      >
                        <span
                          style={
                            Object {
                              "flexGrow": 1,
                              "textAlign": "center",
                            }
                          }
                        >
                          Tietorakenteita luotu
                        </span>
                      </GridColumn>
                      <GridColumn
                        width={3}
                      >
                        <span>
                          undefined p / undefined p
                        </span>
                      </GridColumn>
                    </GridRow>
                  </Grid>
                </FormField>
              </CardContent>
            </Card>
            <Card
              className="checklistCard"
              color="red"
              fluid={true}
              key="Dokumentaatio"
            >
              <CardContent
                header="Dokumentaatio"
              />
              <CardContent
                className="checklistCardRow"
                key="4"
                onClick={[Function]}
              >
                <FormField>
                  <Grid>
                    <GridRow
                      style={
                        Object {
                          "cursor": "pointer",
                          "userSelect": "none",
                        }
                      }
                    >
                      <GridColumn
                        width={3}
                      >
                        <Icon
                          as="i"
                          name="circle outline"
                          size="large"
                          style={
                            Object {
                              "color": "black",
                            }
                          }
                        />
                      </GridColumn>
                      <GridColumn
                        width={10}
                      >
                        <span
                          style={
                            Object {
                              "flexGrow": 1,
                              "textAlign": "center",
                            }
                          }
                        >
                          Readme
                        </span>
                      </GridColumn>
                      <GridColumn
                        width={3}
                      >
                        <span>
                          undefined p / undefined p
                        </span>
                      </GridColumn>
                    </GridRow>
                  </Grid>
                </FormField>
              </CardContent>
              <CardContent
                className="checklistCardRow"
                key="5"
                onClick={[Function]}
              >
                <FormField>
                  <Grid>
                    <GridRow
                      style={
                        Object {
                          "cursor": "pointer",
                          "userSelect": "none",
                        }
                      }
                    >
                      <GridColumn
                        width={3}
                      >
                        <Icon
                          as="i"
                          name="circle outline"
                          size="large"
                          style={
                            Object {
                              "color": "black",
                            }
                          }
                        />
                      </GridColumn>
                      <GridColumn
                        width={10}
                      >
                        <span
                          style={
                            Object {
                              "flexGrow": 1,
                              "textAlign": "center",
                            }
                          }
                        >
                          Tuntikirjanpito
                        </span>
                      </GridColumn>
                      <GridColumn
                        width={3}
                      >
                        <span>
                          undefined p / undefined p
                        </span>
                      </GridColumn>
                    </GridRow>
                  </Grid>
                </FormField>
              </CardContent>
            </Card>
            <div>
              <Form
                as="form"
                className="checklistOutput"
                onSubmit={[Function]}
              >
                <p
                  className="checklistOutputPoints"
                >
                  points: 
                  NaN
                </p>
                <input
                  name="points"
                  type="hidden"
                  value={NaN}
                />
                <Button
                  as="button"
                  type="submit"
                >
                  Copy to review field
                </Button>
              </Form>
            </div>
          </div>
        </GridColumn>
      </GridRow>
    </Grid>
  </div>
</div>
`;
